Feature/input

MarinusRetra opened the pull request:
## Changelog:
- Added ``PlayerInputSystem.cs`` in **/_Scripts/Utility**
- Added ``Utility.asmdef`` in **/_Scripts/Utility**
- Added ``InputManager`` in **/_Scripts/Managers**
- Added ``Managers.asmdef`` in **/_Scripts/Managers**
- Added ``/Controller folder`` in **/_Scripts**
- Added ``PlayerController`` in **/_Scripts/Controller**
- Added ``Controller.asmdef`` in **/_Scripts/Controller**
- Added ``PlayerInputSystemPlayModeTest.cs`` in **/Unit Tests/Playmode**
- Added ``InputScene.unity`` in **/Scenes/Temp**
- Added Following Assembly Definition Reference to **/Unit Tests/Playmode/Playmode.asmdef** ``Managers.asmdef``,``Utility.asmdef``,``UnityInputSystem.asmdef``,``UnityInputSystem.TestFramework.asmdef`` in                
- Changed default action mappings to custom UI and Player action mappings ;

## Known issues
### Slight Input lag
Most Input events in the InputManager are triggered during the performed phase of an InputAction instead of during the slightly faster started phase this is to prevent input events from being called twice per input detection, leading to a 1 frame delay when input is detected to when the fucntion in the PlayerController is called.
## How To Use/Test
- Add InputManager to an empty gameobject in the scene
- Add PlayerController to the desired gameobject
- For every button pressed that is assigned in the InputSystem a message will appear in the console when any of these are pressed ## Unit Tests
## Unit Tests
[SwapActionMapps](https://github.com/user-attachments/assets/c0f201b3-3d64-42db-bae9-12f0388cf012)
![InvokeEventInputTest](https://github.com/user-attachments/assets/5d1cc403-81bd-4b7a-b1ae-01d68744b31f)

NiekMSoftware committed:
Added singleton pattern

NiekMSoftware committed:
Set up InputManager for full dev

MarinusRetra committed:
Enabled GPU Resident Drawer

- Disabled unity splash screen
~ Changed rendering path from forward to forward+
~ set BatchRendererGroup Variants to Keep All.
~ Set GPU Resident Drawer to Instanced Drawing.

NiekMSoftware committed:
Added State and IState

+ IState.cs
+ State.cs

MikeOomen committed:
Add multiplayer packages

+ Multiplayer Center Quickstart Content
+ Multiplayer Play Mode
+ Multiplayer Services
+ Multiplayer Tools
+ Multiplayer Widgets
+ Netcode for GameObjects

NiekMSoftware committed:
Added Machine.cs

+ FSM_SampleScene.Unity
+ Machine.cs

NiekMSoftware committed:
Actually added StateMachine

+ StateMachine.cs, forgot to add it apparently.

NiekMSoftware committed:
Implemented StateMachine

~ StateMachine
~ State, State constructor takes the Machine as reference.

NiekMSoftware committed:
Added Entity

+ Entity.cs
~ StateMachine, Changed the constructor to take an Entity as arg
~ State, Changed to an abstract class

NiekMSoftware committed:
Modified Entity

~ Entity, Renamed attribute Rigidbody to Body

MikeOomen committed:
Added Relay and Lobby package

+ Relay
+ Lobby

NiekMSoftware committed:
Added Unit Testing

Everything works as intended, FiniteStateMachine is left as a module

MikeOomen committed:
Merge pull request #2 from MarinusRetra/feature/fsm

Finite State Machine

MarinusRetra committed:
AddedPlayerManager

+ Singleton instance InputManager reads gameplay and UI input by subscribing to the PlayerInputSyste, events.
+ InputManager calls its own events after reading PlayerInputSystem events
+ PlayerController subscribes to InputManager to call functions based on player input

MarinusRetra committed:
Merge branch 'feature/main' into feature/input

MarinusRetra committed:
Unit Tests Input

+ Test 1 checks if the events are called when input is detected
+ Test 2 checks if the active ActionMap swaps from Player to UI and back when the cancel key is pressed
~ Changed logic inside the functions subscribed to the InputSystem to only run once when a button is pressed

NiekMSoftware reviewed (approved):
# Review
Code looks good and structures, there are some redundant calls inside the `PlayerController`, maybe refactor these to just: `Move`(as an example, you don't have to explicitly say event here anymore).

## Input Manager
In your `InputManager.cs`, I recommend refactored `Player` and `UI` to these named to give off more clarity: `PlayerActions`, `UiActions`. These can also be made `private`, as you won't call these specifically from anywhere.

## Abilities
Perhaps for the Abilities, just an idea: Name them, Primary, Secondary etc. instead of `Ability1` as.

## Comments
As for your comments, make sure to include a space like this: `// tmpl8`. And remove any unnecessary comments such as line `53` in `InputManager.cs`.

Please make sure to apply these changes before merging, overall good job.

MarinusRetra committed:
~Refactorizations InputSystem

~Shortened function names in Player Controller
~Renamed UI and Player in inputmanager to UI_Actions and Player_Actions
~Renamed Abiltities1,2,3 to Primary Secondary and Tertiary
~Removed unnecessary code from the PlayerInputSystemPlaymodeTest script
~ Overall edited and removed unnecessary comments
